snapcraft_parser/main.py

```diff
@@ -0,0 +1,212 @@
+    origin: lp:~joetalbott/+junk/snapcraft-parser-example
+    description: example
+    project-part: main
+    parts: [part1]
+"""
+    return output
+
+
+def _get_origin_data(origin_dir):
+    origin_data = {}
+    with open(os.path.join(origin_dir, 'snapcraft.yaml'), "r") as fp:
+        origin_data = yaml.load(fp)
+
+    return origin_data
+
+
+def run(args):
+    output = _get_output()
+
+    # XXX: This can't remain in memory if the list gets very large, but it
+    # should be okay for now.
+    master_parts_list = {}
+
+    data = yaml.load(output)
+    for key, value in data.items():
+        parts_list = {}
+        # Store all the parts listed in 'after' for each included part so that
+        # we can check later that we aren't missing any parts.
+        # XXX: What do we do about 'after' parts that should be looked for in
+        # the wiki?  They should be in the master parts list.
+
+        after_parts = set()
+        logger.debug("Processing part %s", key)
+        origin = data[key].get("origin")
+        origin_type = data[key].get("origin-type")
+        project_part = data[key].get("project-part")
+        subparts = data[key].get("parts", [])
+
+        if origin is not None:
+            # TODO: this should really be based on the origin uri not
+            # the part name to avoid the situation where there are multiple
+            # parts being pulled from the same repo branch.
+            origin_dir = os.path.join(BASE_DIR, key)
+            try:
+                os.makedirs(origin_dir)
+            except FileExistsError:
+                pass
+
+            class Options():
+                pass
+
+            options = Options()
+            setattr(options, "source", origin)
+
+            if origin_type is not None:
+                setattr(options, "source_type", origin_type)
+
+            sources.get(origin_dir, None, options)
+
+            origin_data = _get_origin_data(origin_dir)
+
+            parts = origin_data.get('parts', {})
+
+            source_part = parts.get(project_part)
+            if source_part is not None:
+                after = source_part.get('after', [])
+
+                if after:
+                    after = _update_after_parts(project_part, after)
+                    after_parts.update(set(after))
+
+                parts_list[project_part] = source_part
+
+            for part in subparts:
+                source_part = parts.get(part)
+                # TODO: get any dependent parts here as well using
+                # 'project_part' to namespace them.
+                if source_part is not None:
+                    parts_list[_namespaced_partname(project_part,
+                                                    part)] = source_part
+
+        # after_parts should already be namespaced at this point.
+
+        if _valid_parts_list(parts_list, after_parts):
+            master_parts_list.update(parts_list)
+
+    _write_gzipped_parts_list(BASE_DIR, master_parts_list)
```

Comment:
It will feel forced until you get used to the right way ;)
I'm joking, of course. When I finish my comments with a question mark, I'm honestly asking what do you think about my proposal. Feel free to disagree. If I have a strong opinion on something, I will let you know.

Personally, I like the function a little better now and if I were the one writing the code, I would try to make even more internal functions. Not a blocker, of course, just saying.